### ISSUE TITLE
P2P - UI/UX Design/Mocks 

### ISSUE THREAD

**Issue by ekigamba** [opensrp/fhircore#465]
1. Is there a new peer-to-peer UI from the one implemented on android-p2p-sync?
2. What are the possible changes?
3. Is the UX change going to be significant?

cc @rowo

**Comment by rowo** [comment]
@ekigamba It sounds like the workflow may change after getting feedback from LMH that people were confused by the two buttons on the initial screen. The change would be having a single "connect" button where the handshake is complete (both sides confirming) — and then offering the user a way to send data. So in this case after a "handshake" is established, would it be possible if the receiving user just had to accept the transfer or even didn't have to do anything further?

**Comment by pld** [comment]
Going to assign to @isabelcshaw to follow-up on UI reqs, we want this to be the same sync UI for all projects

**Comment by ekigamba** [comment]
@rowo I think this will be possible. The users will lose flexibility of choosing whether they want to send or receive the data. Additionally, we'll need to understand the data flow to engineer this into the apps so that it happen automatically

**Comment by rowo** [comment]
@isabelcshaw @ekigamba @pld  We should talk about this since Francis mentioned we should start needing it for the ANC work too (which goes into the LMH work). @HenryRae and I have started some exploration, but they aren't really informed by anything yet. One proposal was to look into how common apps used by CHWs/our clients do file transfer (ODK Collect, LMH Academy app, Android Beam, Nearby Share), but we haven't researched anything yet. 

Next step might be to have a call about this to hear about the new wifi approach and how that might influence things. I'm not sure on a timeline either, so that might be something to think about @isabelcshaw

**Comment by pld** [comment]
Cool yes let's talk, I need to understand the argument for why this can't all happen in the background w/no new UI

**Comment by rowo** [comment]
also i saw some conversation happening on slack.. not sure where that fits in

- With Trevor: https://onaio.slack.com/archives/C01JEDB8WTU/p1631716846053200
- Francis strategy: https://onaio.slack.com/archives/C01MJK0TWAG/p1631541051018300

**Comment by ekigamba** [comment]
@rowo It seems this work is being taken over by the SA team

**Comment by f-odhiambo** [comment]
Latest P2P UX Designs  - https://www.figma.com/proto/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?page-id=5010%3A38742&node-id=5277%3A38741&viewport=364%2C48%2C0.25&scaling=min-zoom&starting-point-node-id=5277%3A38741

**Comment by pld** [comment]
In that UI, on screen 42, when it begins to send data, can the user be able to minimize that and perform other tasks in the app with a UI element showing that sync is in progress?

**Comment by f-odhiambo** [comment]
> In that UI, on screen 42, when it begins to send data, can the user be able to minimize that and perform other tasks in the app with a UI element showing that sync is in progress?

Should we really allow this to happen ? 
If yes, my thought is 
1. if the user should come out of the screen and go to another screen/activity  then alert them once sync is complete  - in app notification 
2. If they come out of the app, go to a different app or lock phone ,  send a system wide notification alert

**Comment by rowo** [comment]
@f-odhiambo @pld how long will the sync process take? This is kind of one of those things that can get hairy if you make changes to a record that's been synced already. The question is more do you trust the app to be able to handle all this stuff going on at the same time and to sync perfectly? We can conform the UI to support what the technology is capable of doing well.

**Comment by pld** [comment]
We'd probably have to disable edits, so then maybe there's no point unblocking the user?

But yea how long those it take is my main question too and I don't know the answer

**Comment by Rkareko** [comment]
> Latest P2P UX Designs - https://www.figma.com/proto/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?page-id=5010%3A38742&node-id=5277%3A38741&viewport=364%2C48%2C0.25&scaling=min-zoom&starting-point-node-id=5277%3A38741

@f-odhiambo @rowo  cc @ekigamba There is need for 
1. An authentication / pair confirmation screen before data transfer starts on screen 27.
2. Add a screen for the receiving device similar to screen 27
3. Because of the request to support CHA >>> CHA sync documented [here](https://github.com/opensrp/fhircore/issues/281#issuecomment-981735626) we may require both the send and receive button on the same page (page 27). This only applies when a CHA connects to a CHA
4. On screen 30 could we say `X number of records sent` instead of specifying the type of records. In most cases the data being transferred is non-demographic  data e.g observations, encounters e.t.c that might not make sense to the user and cannot be termed as patients or families. This is for users to understand that records were sent even if it's only patient history data and not actual patients / families.

**Comment by dubdabasoduba** [comment]
> > Latest P2P UX Designs - https://www.figma.com/proto/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?page-id=5010%3A38742&node-id=5277%3A38741&viewport=364%2C48%2C0.25&scaling=min-zoom&starting-point-node-id=5277%3A38741
> 
> @f-odhiambo @rowo cc @ekigamba There is need for
> 
> 1. An authentication / pair confirmation screen before data transfer starts on screen 27.
> 2. Add a screen for the receiving device similar to screen 27
> 3. Because of the request to support CHA >>> CHA sync documented [here](https://github.com/opensrp/fhircore/issues/281#issuecomment-981735626) we may require both the send and receive button on the same page (page 27). This only applies when a CHA connects to a CHA
> 4. On screen 30 could we say `X number of records sent` instead of specifying the type of records. In most cases the data being transferred is non-demographic  data e.g observations, encounters e.t.c that might not make sense to the user and cannot be termed as patients or families. This is for users to understand that records were sent even if it's only patient history data and not actual patients / families.

@Rkareko @ekigamba Can we make 1 configurable for team members at the same level. @rowo this might also be a complication in the overall UX/UI, if so we can just add the authentication for everyone.

**Comment by ekigamba** [comment]
> > > Latest P2P UX Designs - [figma.com/proto/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?page-id=5010%3A38742&node-id=5277%3A38741&viewport=364%2C48%2C0.25&scaling=min-zoom&starting-point-node-id=5277%3A38741](https://www.figma.com/proto/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?page-id=5010%3A38742&node-id=5277%3A38741&viewport=364%2C48%2C0.25&scaling=min-zoom&starting-point-node-id=5277%3A38741)
> > 
> > 
> > @f-odhiambo @rowo cc @ekigamba There is need for
> > 
> > 1. An authentication / pair confirmation screen before data transfer starts on screen 27.
> > 2. Add a screen for the receiving device similar to screen 27
> > 3. Because of the request to support CHA >>> CHA sync documented [here](https://github.com/opensrp/fhircore/issues/281#issuecomment-981735626) we may require both the send and receive button on the same page (page 27). This only applies when a CHA connects to a CHA
> > 4. On screen 30 could we say `X number of records sent` instead of specifying the type of records. In most cases the data being transferred is non-demographic  data e.g observations, encounters e.t.c that might not make sense to the user and cannot be termed as patients or families. This is for users to understand that records were sent even if it's only patient history data and not actual patients / families.
> 
> @Rkareko @ekigamba Can we make 1 configurable for team members at the same level. @rowo this might also be a complication in the overall UX/UI, if so we can just add the authentication for everyone.

@dubdabasoduba Are you talking about the authentication screen? What do you mean about making one configurable for team members?

The aim of having the authentication/pair confirmation screen is to confirm connection so that devices, CHAs and CHSS, in one room do not forcefully or incorrectly connect to other devices. The device being connected to has to accept the connection request and have a confirmation code to uniquely identify the two connecting devices. The assumption is that the CHW/CHSS name or device name might not be unique enough in some cases.

**Comment by HenryRae** [comment]
I had a call with Richard and Kigamba on this,
This is for reference purposes.
We do have 2 sides(Sender i.e CHA, and receiver i.e CHSS) included in the attached screenshot.

The idea is that 
1. sender(CHA) has their wifi turned on(if not turned on, can we turn it on automatically?), 
2. [scans for nearby devices](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A26863)
3. [chooses a device to send to](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A26897) (e.g James' Phone in the screenshot, assuming James is the CHSS device)
4. at this point on CHSS's device(James Phone) with the app open and wifi turned on, we show a ["A device wants to pair" modal](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A27177), and when CHSS chooses to pair with CHA device
5. now we can show [send data option](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A27002) on CHA's device which when clicked initiates the [data transfer](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A27110).
6. same [progress](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A27207) is shown on the CHSS side once data transfer is initiated by the CHA device.
7. and when done, we show a [data transfer completed](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A27088) message on both devices.

cc: @rowo @f-odhiambo @ekigamba


<img width="1320" alt="Screenshot 2022-02-23 at 16 37 52" src="https://user-images.githubusercontent.com/31955469/155330325-501f1b94-5119-48dc-b485-5bfbd9ff712d.png">

**Comment by pld** [comment]
I don't think we can turn it on automatically, we'll need to show a modal asking them to enable wifi. Also, as @ekigamba pointed out elsewhere, if they haven't syncd GPS recently we may need to wait for GPS to sync so we can verify the time is correct

**Comment by dubdabasoduba** [comment]
> I don't think we can turn it on automatically, we'll need to show a modal asking them to enable wifi. Also, as @ekigamba pointed out elsewhere, if they haven't syncd GPS recently we may need to wait for GPS to sync so we can verify the time is correct

@pld We might want to also inform the user about the GPS sync. @ekigamba @Rkareko can we block the functionality till Wifi is turned on and GPS sync is complete?

**Comment by pld** [comment]
Yes agreed, if either isn't active maybe we can have a modal with the functionalities required, I guess gps is a bit more complicated since it's not just active but synced

**Comment by Rkareko** [comment]
> I had a call with Richard and Kigamba on this, this below is for reference purposes. We do have 2 sides(Sender i.e CHA, and receiver i.e CHSS) included in the attached screenshot.
> 
> The idea is that
> 
> 1. sender(CHA) has their wifi turned on(if not turned on, can we turn it on automatically?),
> 2. [scans for nearby devices](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A26863)
> 3. [chooses a device to send to](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A26897) (e.g James' Phone in the screenshot, assuming James is the CHSS device)
> 4. at this point on CHSS's device(James Phone) with the app open and wifi turned on, we show a ["A device wants to pair" modal](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A27177), and when CHSS chooses to pair with CHA device
> 5. now we can show [send data option](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A27002) on CHA's device which when clicked initiates the [data transfer](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A27110).
> 6. same [progress](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A27207) is shown on the CHSS side once data transfer is initiated by the CHA device.
> 7. and when done, we show a [data transfer completed](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A27088) message on both devices.
> 
> cc: @rowo @f-odhiambo @ekigamba
> 
> <img alt="Screenshot 2022-02-23 at 16 37 52" width="1320" src="https://user-images.githubusercontent.com/31955469/155330325-501f1b94-5119-48dc-b485-5bfbd9ff712d.png">

According to the mock ups the receiver would be listening to connection requests all the time so that the `Pair with device` modal can be called when a connection request comes in. When testing using the [sample p2p wifidirect implementation](https://github.com/opensrp/p2p) we found out that for peer discovery to happen either the wifi-direct listeners (in the app) or the phone wifi-direct settings page have to be running.

In this case, having the wifi-direct Listeners running always would mean that wifi-direct would have to be on and advertising while the app is running irrespective of whether the users will perform P2P sync or not. This will lead to faster battery drain as well as the app using processing power for advertising.

Due to this we propose that both the sender and receiver devices initiate the P2P flow when it is required. Both would be presented with a `scan for devices` screen as shown in the mockup below

![revised_p2p_mockups](https://user-images.githubusercontent.com/47570855/155504027-711aa3f0-1f0d-491c-89fe-a1d0fc50e8ca.png)


@HenryRae @dubdabasoduba

**Comment by HenryRae** [comment]
is it possible to show the system default modal requesting the user to turn on wifi in P2P?
@pld

**Comment by pld** [comment]
> is it possible to show the system default modal requesting the user to turn on wifi in P2P? and also GPS sync
> 
> If yes, I'll update the copy on the first screen. @pld

@dubdabasoduba ^?

**Comment by ekigamba** [comment]
> > I don't think we can turn it on automatically, we'll need to show a modal asking them to enable wifi. Also, as @ekigamba pointed out elsewhere, if they haven't syncd GPS recently we may need to wait for GPS to sync so we can verify the time is correct
> 
> @pld We might want to also inform the user about the GPS sync. @ekigamba @Rkareko can we block the functionality till Wifi is turned on and GPS sync is complete?

- GPS Sync will be required to happen during app start or when the date changes but not required during P2P sync process. The GPS sync helps preserve accuracy of records and consistency across all devices.
- Depending on the Android version, the user might need to turn on wifi-direct and/or wifi manually. We will redirect to the settings page for some devices

**Comment by ekigamba** [comment]
> > I don't think we can turn it on automatically, we'll need to show a modal asking them to enable wifi. Also, as @ekigamba pointed out elsewhere, if they haven't syncd GPS recently we may need to wait for GPS to sync so we can verify the time is correct
> 
> @pld We might want to also inform the user about the GPS sync. @ekigamba @Rkareko can we block the functionality till Wifi is turned on and GPS sync is complete?

- GPS Sync will be required to happen during app start or when the date changes but not required during P2P sync process. The GPS sync helps preserve accuracy of records and consistency across all devices.
- Depending on the Android version, the user might need to turn on wifi-direct and/or wifi manually. We will redirect to the settings page for some devices

**Comment by HenryRae** [comment]
So should we have a design for the worst-case scenario i.e requesting the user to turn on wiFi manually or this too can be a system design/message?

**Comment by dubdabasoduba** [comment]
> So should we have a design for the worst-case scenario i.e requesting the user to turn on wiFi manually or this too can be a system design/message? @ekigamba

I think we should add a modal to ask for Wifi to be turned on immediately the land on the first p2p screen. This should be shown any time wifi or wifi-direct is not turned on.

**Comment by dubdabasoduba** [comment]
> > > I don't think we can turn it on automatically, we'll need to show a modal asking them to enable wifi. Also, as @ekigamba pointed out elsewhere, if they haven't syncd GPS recently we may need to wait for GPS to sync so we can verify the time is correct
> > 
> > 
> > @pld We might want to also inform the user about the GPS sync. @ekigamba @Rkareko can we block the functionality till Wifi is turned on and GPS sync is complete?
> 
> * GPS Sync will be required to happen during app start or when the date changes but not required during P2P sync process. The GPS sync helps preserve accuracy of records and consistency across all devices.
> * Depending on the Android version, the user might need to turn on wifi-direct and/or wifi manually. We will redirect to the settings page for some devices

@ekigamba In the app design do we have a modal asking the user to turn on GPS before login? @HenryRae We might need a screen for that. I think we can add this to a separate issue.

**Comment by dubdabasoduba** [comment]
> > is it possible to show the system default modal requesting the user to turn on wifi in P2P? and also GPS sync
> > If yes, I'll update the copy on the first screen. @pld
> 
> @dubdabasoduba ^?



> > I had a call with Richard and Kigamba on this, this below is for reference purposes. We do have 2 sides(Sender i.e CHA, and receiver i.e CHSS) included in the attached screenshot.
> > The idea is that
> > 
> > 1. sender(CHA) has their wifi turned on(if not turned on, can we turn it on automatically?),
> > 2. [scans for nearby devices](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A26863)
> > 3. [chooses a device to send to](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A26897) (e.g James' Phone in the screenshot, assuming James is the CHSS device)
> > 4. at this point on CHSS's device(James Phone) with the app open and wifi turned on, we show a ["A device wants to pair" modal](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A27177), and when CHSS chooses to pair with CHA device
> > 5. now we can show [send data option](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A27002) on CHA's device which when clicked initiates the [data transfer](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A27110).
> > 6. same [progress](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A27207) is shown on the CHSS side once data transfer is initiated by the CHA device.
> > 7. and when done, we show a [data transfer completed](https://www.figma.com/file/kXJMqeQDyzfXp4urKCfE0R/LMH-eCBIS---focused?node-id=3306%3A27088) message on both devices.
> > 
> > cc: @rowo @f-odhiambo @ekigamba
> > <img alt="Screenshot 2022-02-23 at 16 37 52" width="1320" src="https://user-images.githubusercontent.com/31955469/155330325-501f1b94-5119-48dc-b485-5bfbd9ff712d.png">
> 
> According to the mock ups the receiver would be listening to connection requests all the time so that the `Pair with device` modal can be called when a connection request comes in. When testing using the [sample p2p wifidirect implementation](https://github.com/opensrp/p2p) we found out that for peer discovery to happen either the wifi-direct listeners (in the app) or the phone wifi-direct settings page have to be running.
> 
> In this case, having the wifi-direct Listeners running always would mean that wifi-direct would have to be on and advertising while the app is running irrespective of whether the users will perform P2P sync or not. This will lead to faster battery drain as well as the app using processing power for advertising.
> 
> Due to this we propose that both the sender and receiver devices initiate the P2P flow when it is required. Both would be presented with a `scan for devices` screen as shown in the mockup below
> 
> ![revised_p2p_mockups](https://user-images.githubusercontent.com/47570855/155504027-711aa3f0-1f0d-491c-89fe-a1d0fc50e8ca.png)
> 
> @HenryRae @dubdabasoduba

@Rkareko I think I agree with the plan. We want to preserve both memory and battery on these devices

**Comment by ekigamba** [comment]
@dubdabasoduba
- Yes, we need this https://github.com/opensrp/fhircore/issues/465#issuecomment-1050328505.@HenryRae  This is not for worst-case scenario but it's dependent on the Android version
- We do not have a GPS modal design. [Here is the issue for this](https://github.com/opensrp/fhircore/issues/1083)

**Comment by ekigamba** [comment]
@dubdabasoduba
- Yes, we need this https://github.com/opensrp/fhircore/issues/465#issuecomment-1050328505. @HenryRae  This is not for worst-case scenario but it's dependent on the Android version
- We do not have a GPS modal design. [Here is the issue for this](https://github.com/opensrp/fhircore/issues/1083)

**Comment by Rkareko** [comment]
@HenryRae @dubdabasoduba concerning issue # 3 [here](https://github.com/opensrp/fhircore/issues/465#issuecomment-1044454082) what would you advice as the best way to approach UI/UX ?

**Comment by dubdabasoduba** [comment]
> @HenryRae @dubdabasoduba concerning issue # 3 [here](https://github.com/opensrp/fhircore/issues/465#issuecomment-1044454082) what would you advice as the best way to approach UI/UX ?

- I think we could show both the send button. We however need to emphasize to the client the following items
    - Only 1 device can send data at a time. 
    - Only the person sending data should click send on their device.

**Comment by dubdabasoduba** [comment]
@HenryRae After our meeting, we have the following options
1.  Add a send and receive button to the send initiation screen. This means the receiver will have to click receive for data transfer to be initiated. 
2. Automatically assume the phone that initiates the pairing is the sender and the receiver doesn't have to do anything for them to receive data. 

I would go with option 2 if it's technically viable to track who initiated the pairing. 

@Rkareko @ekigamba .

**Comment by HenryRae** [comment]
@dubdabasoduba would you describe here what exactly does the receiver need to do in order to make the whole process a success for option 2.

**Comment by dubdabasoduba** [comment]
@HenryRae 

**Option 2 Workflow**
---
- The devices search for any available WifiDirect devices. 
- The CHA/CHW clicks and initiates pairing with the other devices. 
   - Here the code will track where the paring request came from. 
   - The phone that requested the pairing will have the send button. 
   - The phone that accepted the pairing won't have any button shown.  The CHA will just wait for the data to be sent.  

cc: @Rkareko @ekigamba

**Comment by HenryRae** [comment]
Thanks @dubdabasoduba,
Just for clarity, does this mean the receiver has to only accept pairing? and do they have to have the app open on any screen or they have to specifically be on the P2P screen as well?

**Comment by dubdabasoduba** [comment]
> Thanks @dubdabasoduba,
> Just for clarity, does this mean the receiver has to only accept pairing? and do they have to have the app open on any screen or they have to specifically be on the P2P screen as well?

- They have to be on the P2P screen.
-  They only have to accept pairing.

**Comment by pld** [comment]
I'm closing this, if there's still WIP to do based on it let's create new specific issues for that
